### PR TITLE
Add basic exception mapping to all REST Data extensions

### LINKED ISSUE #13307: Integrate ORM REST Data Panache with Hibernate Validator
**Description**
I think it would really useful to integrate [the ORM REST Data Panache](https://quarkus.io/guides/rest-data-panache) module with [the Hibernate Validator](https://quarkus.io/guides/validation) extension.

**Implementation ideas**
I've tried to implement a workaround by using the annotations, but I could not make this integration worked.

The modules that I used are:

```
<dependency>
    <groupId>io.quarkus</groupId>
    <artifactId>quarkus-hibernate-orm-rest-data-panache</artifactId>
</dependency>

<dependency>
    <groupId>io.quarkus</groupId>
    <artifactId>quarkus-hibernate-validator</artifactId>
</dependency>
```

My resource:

```
public interface ApplicationResource extends PanacheEntityResource<ApplicationEntity, Long> {
}
```

And my entity:

```
@Entity(name = "application")
public class ApplicationEntity extends PanacheEntity {
    @NotEmpty
    @Column(unique = true, nullable = false)
    public String name;
}
```

When I add an application with an empty name, I got a 500 Internal Server Error because the name is not nullable which is expected, but not nice as I annotated the name field as not empty, therefore I should have received 400 Bad Request.

The only current way to implement this is manually using the validation service, but I would be nice to have something like:

```
public interface ApplicationResource extends PanacheEntityResource<ApplicationEntity, Long> {
    @MethodProperties(validatable = true)
    ApplicationEntity add(ApplicationEntity entity)
}
```

Or even validate the entities *always* only if the hibernate-validator extension does exist in the classpath.

### PR DESCRIPTION
**Pull request by gytis**
Added default exception mappers to the REST Data Panache Hibernate, Mongo and Spring Data REST extensions.

Hibernate:

- `javax.validation.ConstraintViolationException` - `400 Bad Request`
- `org.hibernate.exception.ConstraintViolationException` - `409 Conflict`
- Everything else - `500 Internal Server Error`

Mongo:

- `MongoWriteException` ID E11000 - `409 Conflict`
- Everything else - `500 Internal Server Error`

Spring Data REST

- `javax.validation.ConstraintViolationException` - `400 Bad Request`
- Everything else - `500 Internal Server Error`

Fixes https://github.com/quarkusio/quarkus/issues/13307

### CODE REVIEW

**Comment by geoand** [comment]
Can you please rebase onto master to pick up some CI enhancements?

**Comment by gytis** [comment]
Rebased

**Comment by geoand** [comment]
@loicmathieu want to take a look?

**Comment by gytis** [comment]
Maybe also @Sgitario since he raised the original issue.

**Review by loicmathieu — COMMENTED** [review]
Basically what the exception mapper is doing is climbing down the causes of the exception and decided which message and status code will be written back.
Maybe this can be simplified with a recursive approach on a single method and some if/else chain instead of multiple methods dealing with multiple type of exception ?

File: integration-tests/hibernate-orm-rest-data-panache/pom.xml [hunk 1]
```diff
@@ -25,6 +25,10 @@
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-resteasy-jackson</artifactId>
         </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-hibernate-validator</artifactId>
+        </dependency>
```
Comment (loicmathieu) [inline integration-tests/hibernate-orm-rest-data-panache/pom.xml:31]: You add the dependency but I didn't saw any new tests ?

**Comment by loicmathieu** [comment]
Can you also include duplicate key example inside the test ?

**Comment by gytis** [comment]
@loicmathieu nice tip to use recursion here. I'll change the mappers. 
I'll also add the duplicate ID case to the Mongo integration test. Hibernate already covers that in it's deployment test https://github.com/quarkusio/quarkus/blob/master/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/AbstractPostMethodTest.java#L63 and it doesn't really apply to Spring Data REST because it's persist method treats such POST as an update.

**Comment by gytis** [comment]
@loicmathieu I've updated the PR based on your comment.

**Review by loicmathieu — APPROVED** [review]
LGTM, I do love the recursive throwable cause search implementation ;)

**Comment by gytis** [comment]
It's definitely cleaner. Thanks for the tip!

**Pull request merged**